src/psyclone/tests/psyir/frontend/fparser2_test.py

```diff
@@ -2939,3 +2940,104 @@ def raise_value_error(_1, _2):
     with pytest.raises(ValueError) as err:
         _ = processor.get_routine_schedules("a", ast)
     assert "error to propagate" in str(err.value)
+
+
+def test_structures(fortran_reader, fortran_writer):
+    '''Test that Fparser2Reader parses Fortran types correctly when there
+    is a type declaration with one of the members being initialised,
+    when there is a type declaration with an additional attribute
+    (e.g. one that extends an existing type), when there is a type
+    declaration that contains procedures and when there is a type
+    declaration that both has an additional attribute and contains
+    procedures.
+
+    '''
+    # derived-type with initial value (StructureType)
+    test_code = (
+        "module test_mod\n"
+        "    type, private :: my_type\n"
+        "      integer :: i = 1\n"
+        "      integer :: j\n"
+        "    end type my_type\n"
+        "end module test_mod\n")
+    psyir = fortran_reader.psyir_from_source(test_code)
```

Comment:
This fails in the current implementation, presumably because the parser is being given an empty symbol table. However, it is not as simple as connecting this symbol table with a parent as the structure is being parsed before the variable declarations so they do not yet exist in the symbol table. We could give them an unknown type then try to resolve at the end, or could parse structure types after other variables, although could this cause other problems?.

Reply:
I've found a solution, however it required making a couple of quite heavy changes. 1: I've removed the check for wildcard use statements (or unknownfortrantype) for a variable to be able to be added to a symbol table. 2: I've allowed _process_decln to specialise a symbol into a datasymbol. I'm not sure whether what I've done is the "right" way so am going to give this back for review even though some tests need fixing and possibly new ones creating. In essence what I do is add a Symbol to the symbol table with an unknown type for initial values that are declared outside of a type and then specialise this symbol when its declaration is found.